docs/architecture/adr-065-custom-event-indexing.md

```diff
@@ -0,0 +1,341 @@
+      }
+    }
+  }
+
+  // index end_block events...
+  // execute sqlStmt db query...
+}
+
+func (es *PSQLEventSink) IndexTxEvents(txr *abci.TxResult) error {
+  sqlStmtEvents := sq.Insert("tx_events").Columns("key", "value", "height", "hash", "tx_result_id")
+  sqlStmtTxResult := sq.Insert("tx_results").Columns("tx_result")
+
+
+  // store the tx result
+  txBz, err := proto.Marshal(txr)
+  if err != nil {
+    return err
+  }
+
+  sqlStmtTxResult = sqlStmtTxResult.Values(txBz)
+
+  // execute sqlStmtTxResult db query...
+
+  // index the reserved height and hash indices
+  hash := types.Tx(txr.Tx).Hash()
+  sqlStmtEvents = sqlStmtEvents.Values(types.TxHashKey, hash, txr.Height, hash, txrID)
+  sqlStmtEvents = sqlStmtEvents.Values(types.TxHeightKey, txr.Height, txr.Height, hash, txrID)
+
+  for _, event := range result.Result.Events {
+    // only index events with a non-empty type
+    if len(event.Type) == 0 {
+      continue
+    }
+
+    for _, attr := range event.Attributes {
+      if len(attr.Key) == 0 {
+        continue
+      }
+
+      // index if `index: true` is set
+      compositeTag := fmt.Sprintf("%s.%s", event.Type, string(attr.Key))
+			
+      // ensure event does not conflict with a reserved prefix key
+      if compositeTag == types.TxHashKey || compositeTag == types.TxHeightKey {
+        return fmt.Errorf("event type and attribute key \"%s\" is reserved; please use a different key", compositeTag)
+      }
+		
+      if attr.GetIndex() {
+        sqlStmtEvents = sqlStmtEvents.Values(compositeKey, string(attr.Value), txr.Height, hash, txrID)
+      }
+    }
+  }
+
+  // execute sqlStmtEvents db query...
+}
+
+func (es *PSQLEventSink) SearchBlockEvents(ctx context.Context, q *query.Query) ([]int64, error) {
+  sqlStmt = sq.Select("height").Distinct().From("block_events").Where(q.String())
+
+  // execute sqlStmt db query and scan into integer rows...
+}
+
+func (es *PSQLEventSink) SearchTxEvents(ctx context.Context, q *query.Query) ([]*abci.TxResult, error) {
+  sqlStmt = sq.Select("tx_result_id").Distinct().From("tx_events").Where(q.String())
+
+  // execute sqlStmt db query and scan into integer rows...
+  // query tx_results records and scan into binary slice rows...
+  // decode each row into a TxResult...
+}
+```
+
+### Configuration
+
+The current `tx_index.indexer` configuration would be changed to accept a list
+of supported `EventSink` types instead of a single value.
+
+Example:
+
+```toml
+[tx_index]
+
+indexer = [
+  "kv",
+  "psql"
+]
+```
+
+If the `indexer` list contains the `null` indexer, then no indexers will be used
+regardless of what other values may exist.
+
+Additional configuration parameters might be required depending on what event
+sinks are supplied to `tx_index.indexer`. The `psql` will require an additional
+connection configuration.
+
+```toml
+[tx_index]
+
+indexer = [
+  "kv",
+  "psql"
+]
+
+pqsql_conn = "postgresql://<user>:<password>@<host>:<port>/<db>?<opts>"
```

Comment:
Other than this connection string, do we foresee any other config parameters that we might need?

Reply:
No. As we support more sinks, they'll need their own custom configuration. e.g. a kafka connection string, `kafka_conn`.